### PR TITLE
Removes wayland socket support by default

### PR DESCRIPTION
**Pull request by gdonval**
Wayland is not supported upstream and leaving it here increases the attack surface for no benefit whatsoever. Worse actually, users could believe wayland is supported when it is not.

For "experimental support" (as specified in the comment), user should use `flatseal` or `flatpak override` directly to enable the socket.

### CODE REVIEW

**Comment by gdonval** [comment]
>
> It is an extra socket giving access to non-functional extra software that
would not be accessible otherwise so it mechanically increases the surface
of attack.

We have 2 cases:

* Either this flatpak version provides support for activating the
(unsupported upstream) wayland backend, in which case the socket can stay;
* Or no support is provided by default in which case the socket is
superfluous.

It sounds disingenuous to me to activate a socket by default that cannot be
used by default just because *if it so happens you started Element with a
few mile-long extra command line flags*, it would save you a
`--socket=wayland` flag. (Funnily enough, a flag you would not even have to
type actually since it can trivially be overridden, contrary to command
line arguments).

The whole point is not starting a holy war: it's fine if Flatpak's Element
does not support wayland, it's just that in the same way Element has
nothing to do with reading the host's `/etc` folder, it should not have
access to a superfluous socket. At best it's misleading.

**Comment by Erick555** [comment]
> would not be accessible otherwise so it mechanically increases the surface
> of attack.

It doesn't. Just look at what wayland does and what x11 does. The latter is superset of the former in terms of attack surface. Wayland permission is completely meaningless for security when x11 is available.

> Either this flatpak version provides support for activating the
> (unsupported upstream) wayland backend

Wayland is supported. If it doesn't work then you may report it and upstream should fix it as [other people do](https://github.com/vector-im/element-web/issues/18607)

**Comment by gdonval** [comment]
> would not be accessible otherwise so it mechanically increases the surface
> of attack.
>
> It doesn't. Just look at what wayland does and what x11 does. The latter
> is superset of the former in terms of attack surface. Wayland permission is
> completely meaningless for security when x11 is available.
>

When the wayland socket is not available, I can't launch a GUI using
wayland. Therefore X11 is not a superset but extra.


Either this flatpak version provides support for activating the
> (unsupported upstream) wayland backend
>
> Wayland is supported. If it doesn't work then you may report it and
> upstream should fix it as other people do
> <https://github.com/vector-im/element-web/issues/18607>
>
>
Wayland is not supported by upstream and it's not supported here: when I
launch Element's flatpak under wayland it still uses XWayland. The wayland
socket is therefore unneeded and unused.

I have not only reported but provided a PR to activate Wayland under
wayland session by default here. It got rejected because upstream does not
support it.

You seriously have to sort things out between yourselves: it's one or the
other, not both or neither depending on your mood.

**Comment by Erick555** [comment]
> When the wayland socket is not available, I can't launch a GUI using
> wayland. Therefore X11 is not a superset but extra.

We're talking abut security not functionality. You claim that wayland adds attack surface so explain how or stop spreading false claims.

I proved upstream actively supports wayland. Right now you need enable it manually and that's it.

**Comment by gdonval** [comment]
> We're talking abut security not functionality. 

No, don't arbitrarily reframe what this is about.

* `--socket=wayland` adds support for wayland (i.e. provide access to the compositor if available trough special files and sockets) yet that access is useless with provided configuration: if I disable `--socket=X11` and just enable `--socket=wayland`, flatpak Element crashes. That's a fact.
* Those extra accesses are increased attack surface by the mere fact they are there. That's a fact. Just like providing network or `/` access are. The fact that X11 provides dreadful privilege separation has nothing to do with it: with that state of mind, I could make `/` user-owned and share it in all flatpaks because "X11 is broken anyway so why trying pretending we can be secure..."
* If the extra attack surface is justified for functioning (like network for element), it is a necessary risk. If it is not justified (like `/` access or wayland with default configuration), then it's unnecessary risk and part of the point of containerising. The good security track record of wayland and posix file permissions have nothing to do with restricting by principle as an extra layer of security the day it fails.
* If the user is required to do something, like say, random things  that comes to mind... spawning a program with extra flags to make it work like `flatpak run im.riot.Riot  --enable-features=UseOzonePlatform,WaylandWindowDecorations --ozone-platform=wayland`, they are doing something that isn't default action already. That's not what flatpak provided anymore. They can either run `flatpak run --socket=wayland im.riot.Riot  --enable-features=UseOzonePlatform,WaylandWindowDecorations --ozone-platform=wayland` instead as easily or overriding the configuration to do just this.
* Element, be it in flatpak or as provided by upstream, crashes on a wayland-only session (i.e. *no XWayland*) by default. There is no config file, no Element-specific trigger, just the args passthrough implemented by Electron (by opposition to Element). Therefore Element does not support wayland by default: you should align to that.

The tl;dr of this is: if when clicking on the default flatpak-provided `.desktop` file for Element, there is **no way** the program can use a certain capability granted to it (in this case `--socket=wayland`), therefore said capability should not be provided. It is the same for network, host file system, pulseaudio or any other device: providing access to each of them increases the attack surface by making something that should not be accessible, accessible.

BTW, Element is the only flatpak out of the 53 I have that advertises wayland support and yet fails when started in wayland-only mode. That's atop of refusing making it work or stopping advertising for something that simply is not provided.

**Comment by Erick555** [comment]
> No, don't arbitrarily reframe what this is about.

This is exactly what you do for whole discussion. You can't explain what attack surface wayland permission adds in this flatpak on top of x11. It doesn't matter what you think is a fact where there is nothing supporting it.

Claim that wayland doesn't work when isn't enabled is simply ridiculous. You need to convince people running this app with wayland that it doesn't support wayland which I'm afraid is hopeless task.

**Comment by gdonval** [comment]
Answer these two closed questions for me:

1. Does activating `--socket=wayland` expose more of the underlying system to Element?
2. Assuming I am running a wayland compositor, does clicking on the default provided flatpak's `.desktop` icon run Element on wayland?

**Comment by Erick555** [comment]
> Does activating --socket=wayland expose more of the underlying system to Element?

No. Every possible wayland system exposition is exposed through x11 socket. It's like exposing whole `/` then claim exposing `/aaa/bbb/ccc/ddd/fff/ggg/hhh` adds attack surface.

> Assuming I am running a wayland compositor, does clicking on the default provided flatpak's .desktop icon run Element on wayland using its wayland backend (not the xwayland one that only requires X11)?

No. Using native wayland in chromium framework requires passing an optional argument. This is even advertised on flatpak [frontpage](https://flathub.org/apps/details/im.riot.Riot). It works like that on every linux platform and for every chromium/electron app. The question is then why flatpak should block functionality that exist in app? Why flatpak users would need to learn about permissions and overrides only because they want to use built-in app functionality? Since when flatpak is deciding what app feature users are allowed to use and what they aren't?

There are plenty of electron apps on flathub that have wayland permission. Chromium and chrome browsers have it to. I don't know why you opened crusade against this specific app.

**Comment by LorbusChris** [comment]
Please don't do this.

Removing the wayland socket would break HiDPI for anyone passing the flags manually.

If anything, it should be considered to pass the following flags by default:
`--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations`

**Comment by gdonval** [comment]
> No. Every possible wayland system exposition is exposed through x11 socket. It's like exposing whole `/` then claim exposing `/aaa/bbb/ccc/ddd/fff/ggg/hhh` adds attack surface.

Then why does wayland not work when X11 is the only flag set? Or more importantly, why do you care so much about removing explicit `/aaa/bbb/ccc/ddd/fff/ggg/hhh` when `/` is already exposed? Either way, `--socket=wayland` is unnecessary and advertises the wrong thing.  




> No. Using native wayland in chromium framework requires passing an optional argument. 

Good. Then if you go to such length to get proper `wayland` support, you can activate it when necessary, either as a one-off thing or on the same command line that **does** activate wayland support.

Even better, according to you, this is even not necessary at all since `--socket=x11` already provides everything needed without change.

 
> Since when flatpak is deciding what app feature users are allowed to use and what they aren't?

Flatpaks are providing defaults. Only that, defaults. You have a hard time understanding that enabling actual `wayland` support requires active action on the user part to change those defaults. All that at the command line level. The argument then is that said users should activate the socket the default conf does not use (and probably deactivate the x11 one).

BTW, with this justification, you should just expose `host` and let users override it.


> There are plenty of electron apps on flathub that have wayland permission. Chromium and chrome browsers have it to. I don't know why you opened crusade against this specific app.

At the very least `chrome` reads config files so while it's actually not OK there either there is a persistent way (which stays between installs as long as you keep `.var/flatpak` untouched) to activate the wayland backend. But you are right that it is wrong.

**Comment by gdonval** [comment]
> Please don't do this.

We are talking about defaults here. `flatpak --socket=wayland --nosocket=x11 im.riot.Riot` or `flatseal` (if you want a graphical interface) is all that's needed to swap to the correct configuration permanently.

Alternatively, since you already have to type 

```flatpak run im.riot.Riot --ozone-platform-hint=auto --enable-features=WaylandWindowDecorations```

you can use

```flatpak run --socket=wayland im.riot.Riot --ozone-platform-hint=auto --enable-features=WaylandWindowDecorations```

and get exactly the same thing.

> If anything, it should be considered to pass the following flags by default

I tried doing it in PR https://github.com/flathub/im.riot.Riot/pull/301. So that the wayland backend would be used if available (and in case it is problematic somewhere, it could be gated behind an environment variable that users could `flatpak override`). It got rejected because they want to stay as "vanilla" as possible since users who want wayland can clutter their own computer with `.desktop` files and aliases. Since the default is this, flatpak's default should reflect it instead of activating extra bits that aren't used in normal operations.

**Comment by Erick555** [comment]
Instead of asking people to do overrides do it yourself and stop wasting time for this.

**Comment by gdonval** [comment]
You are correct for once.

You are providing dreadful UX. It's appalling to tell people that if they
want to use wayland, they have to delete a desktop file in /usr and create
a replacement one with all the necessary flags somewhere else.

You are already wrapping Element in a script, activating the flags if env
var is present is not rocket science and would solve the problem neatly
withing flatpak while leaving the end decision to the end user and without
disrupting anyone who doesn't care.

At the same time you apparently have no idea what flatpak does or doesn't
and have no clue about how security by capability restriction works. So you
insist exposing something you do not use.

So you are correct: no point talking to someone who's visibly incompetent
from end to end.

**Comment by SISheogorath** [comment]
And I guess with those very conclusive words, that are not very uplifting or particularly nice, I want to end this conversation here. I appreciate your interest and passion regarding the current wayland configuration and how it should be adjusted.

I don't think this conversation is leading anywhere. One reason to keep the wayland socket around for now, is the simple fact that we have added it, to ease the usage of wayland, if people are willing to configure Element to use it, and there you are correct, it would be just one more parameter to add at the right position, however, removing it now, would break a bunch of setups tomorrow and the risks posed by the wayland socket (which are not entirely the same as the x11 socket, since wayland just on a technical level is not just a subset of X11), are ones that can be tolerated.

As mentioned in the other PR, I would recommend to work with upstream towards wayland by default, because it'll benefit everyone and not just Flatpak users. I'm looking forward to hear great news regarding upstream wayland support in the future and wish you the best for now. :)
